### PR TITLE
Make capture groups available inside hooks of regular expression fields.

### PR DESCRIPTION
**Pull request by rsmmr**

### CODE REVIEW

**Comment by rsmmr** [comment]
@bbannier I'm afraid this will be a bit hard to review because it fixes/tweaks some pretty low-level regex stuff, don't worry to much.

File: tests/spicy/types/regexp/parse-ctor-captures-no-sub.spicy [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    x: /abc/ {
+        print $1; # out of range
+        }
+};
+
+public type Y = unit {
+    y: /(a.c)(de*f)(h.j)/ &nosub {
+        print $1; # out of range
+        }
+};
```
Comment (bbannier) [inline tests/spicy/types/regexp/parse-ctor-captures-no-sub.spicy:20]: Let's add a test where we do have capture groups, but the index is out of bounds, e.g.,

```.ruby
public type Y = unit {
    y: /(a.c)(de*f)(h.j)/ {
        print $99; # out of range
        }
};
```

File: doc/programming/parsing.rst [hunk 2]
```diff
@@ -854,11 +854,25 @@ where ``T`` is one of :ref:`spicy_RealType`.
 Regular Expression
 ^^^^^^^^^^^^^^^^^^
 
-When parsing a field through a :ref:`type_regexp` , the expression is
+When parsing a field through a :ref:`type_regexp`, the expression is
```
Comment (bbannier) [inline doc/programming/parsing.rst:857]: We don't currently mention `&anchor` anywhere in the docs.

File: hilti/runtime/src/tests/regexp.cc [hunk 3]
```diff
@@ -118,6 +167,116 @@ TEST_CASE("advance") {
...
+
+    CHECK_EQ(ms_anchor_no_sub_1.advance("Xa"_b, false), std::make_tuple(0, 0));
+    CHECK_EQ(ms_anchor_no_sub_1.captures(Stream("XabbbcdefgX"_b)), Vector<Bytes>());
+
+    CHECK_EQ(ms_anchor_no_sub_2.advance("a"_b, false), std::make_tuple(-1, 1));
+    CHECK_EQ(ms_anchor_no_sub_2.advance("bb"_b, false), std::make_tuple(-1, 2));
+    CHECK_EQ(ms_anchor_no_sub_2.advance("bc"_b, false), std::make_tuple(-1, 2));
+    CHECK_EQ(ms_anchor_no_sub_2.advance("de"_b, false), std::make_tuple(-1, 2));
+    CHECK_EQ(ms_anchor_no_sub_2.advance("fgX"_b, true), std::make_tuple(1, 2));
+    CHECK_EQ(ms_anchor_no_sub_1.captures(Stream("XabbbcdefgX"_b)), Vector<Bytes>());
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/runtime/src/tests/regexp.cc:231]: not sure why?

File: hilti/toolchain/include/ast/ctors/regexp.h [hunk 4]
```diff
@@ -23,7 +23,29 @@ class RegExp : public NodeBase, public hilti::trait::isCtor {
...
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns true if matching of this pattern should be implicitly anchored.
+     */
+    auto isAnchor() const {
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/toolchain/include/ast/ctors/regexp.h:46]: Fixed them all.

File: spicy/toolchain/src/compiler/codegen/parser-builder.cc [hunk 5]
```diff
@@ -1192,7 +1212,18 @@ void ParserBuilder::newValueForField(const type::unit::item::Field& field, const
     }
 
     beforeHook();
-    builder()->addMemberCall(state().self, ID(fmt("__on_%s", field.id().local())), {nvalue}, field.meta());
+
+    std::vector<Expression> args = {nvalue};
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply spicy/toolchain/src/compiler/codegen/parser-builder.cc:1216]: Move should be fine, fixed.

File: hilti/toolchain/include/ast/ctors/regexp.h [hunk 6]
```diff
@@ -23,7 +23,29 @@ class RegExp : public NodeBase, public hilti::trait::isCtor {
...
+     * groups. That's the case if either the regexp has been explicitly marked as
+     * ``&nosub``, or if there are no groups being used.
+     */
+    auto isNoSub() const {
+        if ( AttributeSet::find(attributes(), "&nosub") )
+            return true;
+
+        for ( const auto& p : _patterns ) {
+            if ( auto i = p.find('('); i >= 0 && (i == 0 || p[i-1] != '\\') )
+                return false;
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/toolchain/include/ast/ctors/regexp.h:37]: It's ok to use parentheses even with `&nosub`, one just won't be able to extract the content for the subexpressions after a match.

File: hilti/runtime/src/tests/regexp.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/runtime/src/tests/regexp.cc:231]: Oh, the `_1`. That's a typo.

**Review by bbannier — DISMISSED** [review]

**Comment by rsmmr** [comment]
This got some additional commits at the end fixing Zeek plugin problems I noticed when testing the packages.

**Review by bbannier — APPROVED** [review]

**Pull request merged**